pyresample/geometry.py

```diff
@@ -713,43 +715,301 @@ def copy(self):
+        # - If input lats/lons are xr.DataArray, the specified dims ['x','y'] are ignored
+        src_lons, src_lons_format = _convert_2D_array(self.lons, to='DataArray_Dask', dims=['y', 'x'])
+        src_lats, src_lats_format = _convert_2D_array(self.lats, to='DataArray_Dask', dims=['y', 'x'])
+
+        # Conversion to Geocentric Cartesian (x,y,z) CRS
+        res = da.map_blocks(self._do_transform, latlong, geocent,
+                            src_lons.data,
+                            src_lats.data,
+                            da.zeros_like(src_lons),  # altitude
+                            new_axis=[2],
+                            chunks=(src_lons.chunks[0], src_lons.chunks[1], 3))
+        res = xr.DataArray(res, dims=['y', 'x', 'xyz'])
+
+        # Retrieve new centroids
+        # TODO: make it dask compatible using _upsample_centroid_dask [HELP WANTED]
+        # res1 = da.apply_along_axis(_upsample_centroid_dask,
+        #                            2,
+        #                            res.data,
+        #                            x,
+        #                            y)
+        # res1 = xr.DataArray(res1, dims=['y', 'x', 'coord'], coords=src_lons.coords)
+        res = np.stack(upsample_centroids(res[:, :, 0].data,
+                                          res[:, :, 1].data,
+                                          res[:, :, 2].data, x=x, y=y), axis=2)
+        new_centroids = xr.DataArray(da.from_array(res), dims=['y', 'x', 'xyz'])
+
+        # Back-conversion to geographic CRS
+        lonlatalt = da.map_blocks(self._do_transform, geocent, latlong,
+                                  new_centroids[:, :, 0].data,  # x
+                                  new_centroids[:, :, 1].data,  # y
+                                  new_centroids[:, :, 2].data,  # z
+                                  new_axis=[2],
+                                  chunks=new_centroids.data.chunks)
+
+        # Back-conversion array as input format
+        lons, _ = _convert_2D_array(lonlatalt[:, :, 0], to=src_lons_format, dims=src_lons.dims)
+        lats, _ = _convert_2D_array(lonlatalt[:, :, 1], to=src_lats_format, dims=src_lats.dims)
+
+        # Add additional info if the source array is a DataArray
+        if isinstance(self.lats, xr.DataArray) and isinstance(self.lons, xr.DataArray):
+            lats.attrs = self.lats.attrs.copy()
+            lons.attrs = self.lons.attrs.copy()
+            try:
+                resolution = lons.attrs['resolution'] / ((x + y) / 2)
+                lons.attrs['resolution'] = resolution
+                lats.attrs['resolution'] = resolution
+            except KeyError:
+                pass
+
+        # Return the downsampled swath definition
         return SwathDefinition(lons, lats)
 
+    def extend(self, left=0, right=0, bottom=0, top=0):
+        """Extend the SwathDefinition of n pixels on specific boundary sides.
+
+        By default, it does not extend on any side.
+        """
```

Comment:
You may have had this discussion with Martin in previous comment sections, but what is this method used for? What is the real world use case?

Reply:
```suggestion
    def extend(self, left=0, right=0, bottom=0, top=0):
        """Extend the SwathDefinition by n pixels on specific boundary sides.

        By default, it does not extend on any side.
        """
```

Reply:
I am currently implementing a pipeline for collocating and remapping GEO imagery on LEO sensor swath.
It's very useful to actually have GEO imagery outside the original LEO swath for downstream image processing/ML applications.

As an example, having a LEO sensor with 5 km footprint resolution, I first extend the LEO SwathDefinition of N pixels on both sides along-track, then I upsample the SwathDefinition by a factor of 5 to match the 1km resolution of GEO imagery.  --> `swathdef.extend(top=N, bottom=N).upsample(x=5,y=5) `

Reply:
I stupidly mentioned that I am extending along-track while I am extending the SwathDefinition cross-track ;)

I essentially want to have neighbor information around the swath and upscale it to: 
1. not lose potential fine-scale information available from GEO data (i.e. overshooting tops, inhomogeneity/non-uniform beam filling degree,...). 
2. and if I apply a series of convolutions (i.e. with CNN), the image at the borders gets shrunk when not applying padding, so this is another reason to extend the swath cross-track so that after a series of convolutions/pooling operations I can be aligned with original LEO SwathDefinition.

Reply:
Does the ML code handle the geolocation of the data at all? Or are you telling it (in some way) that the top-left 2x2 pixels of the resampled GEO pixels represent pixel [0, 0] of the LEO data? Or I guess in this extended case the extrapolated pixels of the GEO data don't have any corresponding LEO pixel.

If the ML has a concept of the location of the data, then why not simply slice the GEO data to the bounds of the LEO data instead of resampling?

Note: I'm not trying to argue against this `extend` method I just really want to understand how it is used so I can argue for/against changes in the future.